crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3553,35 +3467,9 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
-    /// Adds a new diagnostic.
-    ///
-    /// The diagnostic does not get added if the rule isn't enabled for this file.
-    pub(super) fn add_diagnostic(
-        &mut self,
-        node: AnyNodeRef,
-        rule: &str,
-        message: std::fmt::Arguments,
-    ) {
-        if !self.db.is_file_open(self.file) {
-            return;
-        }
-
-        // TODO: Don't emit the diagnostic if:
-        // * The enclosing node contains any syntax errors
-        // * The rule is disabled for this file. We probably want to introduce a new query that
-        //   returns a rule selector for a given file that respects the package's settings,
-        //   any global pragma comments in the file, and any per-file-ignores.
-
-        self.types.diagnostics.push(TypeCheckDiagnostic {
-            file: self.file,
-            rule: rule.to_string(),
-            message: message.to_string(),
-            range: node.range(),
-        });
-    }
-
     pub(super) fn finish(mut self) -> TypeInference<'db> {
         self.infer_region();
+        self.types.diagnostics.extend(self.diagnostics.finish());
```

Comment:
I think what's different now is that `extend` adds the diagnostics to `self.types.diagnostics` and `finish` merges the two. 

We should not write to `types.diagnostics` and instead push all diagnostics to the `DiagnosticsBuilder` (we otherwise risk allocating more vecs)
```suggestion
        self.types.diagnostics. = self.diagnostics.finish();
```

Reply:
I don't think we can directly override the `self.types.diagnostics` because otherwise it'll drop the existing diagnostics that's present. This happens because the inference builder is mutually recursive.

Did you mean the following?
```rs
self.diagnostics.extend(self.diagnostics.finish());
```

Reply:
types.diagnostics should be empty at  this point. If it isn't (e.g because of extend), then these code paths should be updated as well to push the diagnostics to the builder instead. Pushing some diagnostics to the builder and some to types.diagnostics is the source of the reordering

Reply:
We could argue that the reordering doesn't matter and we should instead sort the diagnostics but it still has the disadvantage that we allocate one extra vec when the diagnostics are not empty

Reply:
Oh ok, got it. I see the issue.